### PR TITLE
Feature/start with last opened tabs

### PR DESCRIPTION
**Pull request by PoisonousJohn**
- Added "Continue where you left off" setting to the settings page.
- Added "Session" entity to save session-related data to separate file
- Added list of tab states to session.
- Implemented restoration of tabs when opening first browser window

### CODE REVIEW

**Comment by PoisonousJohn** [comment]
This is for #50

**Comment by timsueberkrueb** [comment]
@JohnPoison thank you! I will hopefully be able to look into it tomorrow (or on Friday).

**Comment by PoisonousJohn** [comment]
Also I'm currently looking into saving tab's scrolling progress upon exit. It seems that for this task we'll have to use WebChannel. Also this will give us more flexible way to communicate QML <--> WebPage content

**Comment by timsueberkrueb** [comment]
@JohnPoison That would be cool :) If we have to use WebChannels, we should merge `feature/qtwebengine` first. But couldn't we just get the scroll position by running some JavaScript in the page as we do for theme color support?

**Comment by PoisonousJohn** [comment]
@tim-sueberkrueb sure we can. But isn't that lot easier to modify property of a js object, instead of messing with runJavaScript callbacks? Besides it's not so easy to use runJavaScript callback, since I'm saving sesison data on window close. So I should block closing until callback returns.

I see that web channel can give us a handy way to communicate with web page content. Well, we can even do this in a separate PR

**Comment by timsueberkrueb** [comment]
@JohnPoison I see, that's right. We could also use WebChannels for theme color handling then.

**Comment by timsueberkrueb** [comment]
`feature/qtwebengine` was merged with develop (including the travis fix). You'll probably want to rebase your work on develop. Thanks!

**Comment by PoisonousJohn** [comment]
Lol, it seems i've messed up with git. Will fix it 10th of May.

**Comment by PoisonousJohn** [comment]
@tim-sueberkrueb rebase done.

File: src/ui/tabview/content/SettingsContent.qml [hunk 1]
```diff
@@ -61,11 +61,35 @@ TabContent {
                 }
 
                 TitleLabel {
-                    text: "Start"
+                    text: "Startup"
```
Comment (timsueberkrueb) [inline src/ui/tabview/content/SettingsContent.qml]: Please use `qsTr()` now

File: src/ui/tabview/content/SettingsContent.qml [hunk 2]
```diff
@@ -61,11 +61,35 @@ TabContent {
...
+                }
+
+                Column {
+                    Label { text: "Start new window with" }
```
Comment (timsueberkrueb) [inline src/ui/tabview/content/SettingsContent.qml]: Please use `qsTr()` now

**Review by timsueberkrueb — CHANGES_REQUESTED** [review]
Hey, I'm so sorry that I forgot about this PR ._.
Can you rebase it on `develop`?
Looks good otherwise :+1:
We might want to rethink settings (in general) though (but that's something for v0.5.0 ;)).

File: src/core/session/session.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * $END_LICENSE$
+*/
+
+#include "Session.h"
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: I think that's `"session.h"`.

**Comment by timsueberkrueb** [comment]
I think one thing to consider is whether it would be better if restored tabs would only load on demand (e.g. when the user activates them) as it is the case in Firefox. That would prevent long loading times when many tabs are stored.
What do you think?

**Comment by ShalokShalom** [comment]
Completely agree. +1 

Just think about cases with 20-40 tabs open and more..

**Comment by timsueberkrueb** [comment]
Another problem with this I haven't thought of yet is multi window support.

**Comment by PoisonousJohn** [comment]
Rebased on develop.
Added lazy loading of the tabs.

Regarding multi-window support, right now the session is being saved for the last closed window. I think if we need multi-window support, it should be done as a separate issue.

File: src/ui/tabview/TabController.qml [hunk 4]
```diff
@@ -50,11 +50,24 @@ QtObject {
...
-    function openUrl(url, background) {
-        addTab(TabType.fromUrl(url), {
-            url: url,
-            background: background
-        });
+    function completeTabsRestore()
+    {
+        for (var i = 0; i < tabContentView.pages.count; ++i)
+        {
+            var tabPage = tabContentView.pages.get(i).item;
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabController.qml]: Let's stay consistent with the rest of the JavaScript code and keep the opening curly brace in the same line ;)
```js
for (var i = 0; i < tabContentView.pages.count; ++i) {
```

File: src/ui/tabview/TabController.qml [hunk 5]
```diff
@@ -50,11 +50,24 @@ QtObject {
     signal fullScreenRequested(var request)
     signal openUrlInNewPrivateWindowRequested(var url)
 
-    function openUrl(url, background) {
-        addTab(TabType.fromUrl(url), {
-            url: url,
-            background: background
-        });
+    function completeTabsRestore()
+    {
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabController.qml]: Let's keep the opening curly brace in the same line for consistency ;)
```javascript
function completeTabsRestore() {
```

File: src/ui/tabview/TabPage.qml [hunk 6]
```diff
@@ -75,24 +84,38 @@ Item {
...
-                );
-                break;
-            case TabType.settings:
-                newContent = settingsContentComponent.createObject(
-                    contentContainer,
-                    data.properties
-                );
-                break;
+        if (!loadContent)
+        {
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml]: Let's keep the opening curly brace in the same line for consistency.

File: src/ui/tabview/TabPage.qml [hunk 7]
```diff
@@ -75,24 +84,38 @@ Item {
...
+            // creating 'lazy' tab
+            var lazyProperties = {
+                "initProperties" : data
+            };
+
+            newContent = lazyWebContentComponent.createObject(contentContainer, lazyProperties)
         }
+        else
+        {
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml]: Let's keep the opening curly brace in the same line for consistency.

File: src/ui/tabview/content/LazyWebContent.qml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        property: "iconUrl"
+        value: initProperties['iconUrl']
+        when: lazyCreateConnection.enabled
+    }
+
+    Connections {
+        id: lazyCreateConnection
+
+        function tryCreateContent()
+        {
```
Comment (timsueberkrueb) [inline src/ui/tabview/content/LazyWebContent.qml]: Let's keep the opening curly brace in the same line for consistency.

File: src/ui/tabview/content/LazyWebContent.qml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            if (parentTabPage.tab !== parentTabPage.activeTab)
+                return;
+
+            if (parentTabPage.loadContent)
+            {
```
Comment (timsueberkrueb) [inline src/ui/tabview/content/LazyWebContent.qml]: Let's keep the opening curly brace in the same line for consistency.

File: src/core/session/session.cpp [hunk 10]
```diff
@@ -87,8 +89,16 @@ void Session::load()
...
         auto state = new TabState(this);
         state->setUrl(tabObj["url"].toString());
+        state->setTitle(tabObj["title"].toString());
+        state->setIcon(tabObj["icon"].toString());
         m_tabs.append(state);
     }
+
+    m_activeTab = std::max(m_tabs.count() - 1, 0);
+    if (root.find("activeTab") != root.end())
+    {
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: Let's keep the opening curly brace in the same line for consistency-

File: src/core/session/session.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    Q_OBJECT
+public:
+    explicit Session(QObject *parent = 0);
+
+    Q_INVOKABLE void save(TabsModel* tabs);
+    Q_INVOKABLE QVariantList getTabsToRestore();
+    Q_PROPERTY(int activeTab READ getActiveTab)
+
+    int getActiveTab() { return m_activeTab; }
+signals:
```
Comment (timsueberkrueb) [inline src/core/session/session.h]: We can remove the `signals` and `public slots` sections as they are not used.

**Review by timsueberkrueb — COMMENTED** [review]
Thanks a lot for getting back to this!
Looking good, will check it out and test myself asap.

**Review by timsueberkrueb — COMMENTED** [review]
Looks like the data directory (`/home/tim/.local/share/liri-browser` in my case) is not being created:
`Couldn't open session file for write!`

We should make sure the directory gets created, see https://github.com/lirios/browser/blob/develop/src/core/settings/settings.cpp#L40

**Review by timsueberkrueb — COMMENTED** [review]
One bug I noticed is that it doesn't restore `liri://` tabs. For example, when restarting the browser having a `liri://settings` tab opened, it won't reload. A `This site can’t be reached` page shows up instead.

**Comment by timsueberkrueb** [comment]
Thinking about it, couldn't we merge `LazyWebContent` and `WebContent`? I'm thinking about making `WebContent` just contain the loader wrapping the `WebEngineView` and getting an additional `lazy` property to toggle default or lazy loading.

File: src/core/session/tabstate.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+QString TabState::icon() const
+{
+    return m_icon;
+}
+
+float TabState::readingProgress() const
+{
+    return m_readingProgress;
```
Comment (timsueberkrueb) [inline src/core/session/tabstate.cpp]: Should we remove readingProgress while it's not being used, yet?

File: src/core/session/session.cpp [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    QJsonArray tabsArray;
+
+    for (int i = 0; i < tabs->count(); ++i) {
+        QJsonObject tabObject;
+        auto tab = tabs->get(i);
+        tabObject["url"] = tab->url().toString();
+        tabObject["title"] = tab->title();
+        tabObject["icon"] = tab->iconUrl().toString().replace("image://favicon/","");
+        tabObject["readingProgress"] = 0.f;
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: The JSON fields for settings are currently all `snake_case`. I do prefer `camelCase` though. We can change the settings schema to use `camelCase` (which shouldn't be a problem right now as we don't have a release yet).
Any objections?

File: src/core/session/tabstate.cpp (hunk 12 above)
Comment (PoisonousJohn) [reply src/core/session/tabstate.cpp]: Yep, sure

File: src/core/session/session.cpp (hunk 13 above)
Comment (PoisonousJohn) [reply src/core/session/session.cpp]: No objections. Should I leave it like that for now?

File: src/core/session/session.h (hunk 11 above)
Comment (PoisonousJohn) [reply src/core/session/session.h]: sure

File: src/core/session/session.cpp (hunk 10 above)
Comment (PoisonousJohn) [reply src/core/session/session.cpp]: sorry for that braces, my old habit from c# :)

**Comment by PoisonousJohn** [comment]
> Thinking about it, couldn't we merge LazyWebContent and WebContent? I'm thinking about making WebContent just contain the loader wrapping the WebEngineView and getting an additional lazy property to toggle default or lazy loading.

I have thought about it, and even tried, but something went wrong. Let me try again.
Then it should fix problems with the settings page as well

**Comment by PoisonousJohn** [comment]
> Looks like the data directory (/home/tim/.local/share/liri-browser in my case) is not being created:
> Couldn't open session file for write!
> We should make sure the directory gets created, see https://github.com/lirios/browser/blob/develop/src/core/settings/settings.cpp#L40

I almost sure that I have fixed it. It seems I've lost the change during problems with rebase. Let me fix it again

**Comment by ShalokShalom** [comment]
Oh rebase ^-^

**Comment by ShalokShalom** [comment]
https://hackernoon.com/dont-fear-the-rebase-bca683888dae

**Comment by PoisonousJohn** [comment]
This iteration looks more appropriate. liri-settings bug is fixed as well

File: src/core/session/session.cpp (hunk 10 above)
Comment (timsueberkrueb) [reply src/core/session/session.cpp]: No problem, thanks! :)

**Comment by timsueberkrueb** [comment]
Awesome, thank you! Will check it out a bit later today!

File: src/ui/tabview/TabPage.qml [hunk 14]
```diff
@@ -59,6 +55,15 @@ Item {
     property var contentItem
     property int tabType
 
+    function tryCreateContent() {
+        console.log("Try create content");
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml]: Do you want to keep this logging?

File: src/core/session/session.cpp [hunk 15]
```diff
@@ -42,6 +42,11 @@ Session::Session(QObject *parent)
 
 void Session::save(TabsModel* tabs)
 {
+    if (!QDir(Paths::DataLocation).exists()) {
+        qDebug() << "DataLocation path doesn't exist.";
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: Let's make this a bit more human-friendly, like "App data path doesn't exist".

File: src/ui/tabview/TabPage.qml [hunk 16]
```diff
@@ -69,56 +74,82 @@ Item {
...
+
+        if (loadContent) {
+            contentFabric();
+        }
 
         return true;
     }
 
+    onActiveTabChanged: {
+        console.log("onActiveTabChanged");
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml]: Can we remove this debug statement?

File: src/core/session/session.cpp [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    file.close();
+
+    qDebug() << "Session written to" << Paths::SessionDataFile;
+}
+
+QVariantList Session::getTabsToRestore()
+{
+    QVariantList tabs;
+    for (TabState* state : m_tabs)
+    {
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: Found another curly brace :)

File: src/core/session/session.cpp (hunk 13 above)
Comment (tim-sueberkrueb) [reply src/core/session/session.cpp]: Yep.

File: src/ui/tabview/TabPage.qml [hunk 18]
```diff
@@ -27,11 +27,16 @@ import ".."
 Item {
     id: tabPage
 
+    // binding. should be set from outside
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml:30]: Is this comment still relevant? I don't understand what it means.

File: src/ui/tabview/TabPage.qml [hunk 19]
```diff
@@ -50,6 +55,15 @@ Item {
     property var contentItem
     property int tabType
 
+    function tryCreateContent() {
+        console.log("Try create content");
+        if (contentItem || !loadContent || tab !== activeTab)
+            return;
+
+        console.log("Creating content");
```
Comment (timsueberkrueb) [inline src/ui/tabview/TabPage.qml]: Do we need this debug statement?

**Review by timsueberkrueb — CHANGES_REQUESTED** [review]
Works great now, thank you! One thing I noticed is that you store the icon url but the icons will only show up when the tab fully loads. We either should make the icons load or drop the property.
We also probably want to prevent saving the icon url for `liri://` tabs:
```json
"icon": "qrc:/liri.io/imports/Fluid/Controls/icons/action/settings.svg",
```
It also looks like the icon information gets lost when I restart the browser, tabs get restored but I don't activate them and restart the browser again:
```json
"icon": ""
```
Besides of that, looking good!

File: src/core/session/session.cpp (hunk 13 above)
Comment (timsueberkrueb) [reply src/core/session/session.cpp]: Yep, let's :)

File: src/core/session/session.cpp [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+    QJsonArray tabsArray;
+
+    for (int i = 0; i < tabs->count(); ++i) {
+        QJsonObject tabObject;
+        auto tab = tabs->get(i);
+        tabObject["url"] = tab->url().toString();
+        tabObject["title"] = tab->title();
+        tabObject["icon"] = tab->iconUrl().toString().replace("image://favicon/","");
+        tabObject["readingProgress"] = 0.f;
```
Comment (timsueberkrueb) [inline src/core/session/session.cpp]: `readingProgress` can be removed here, right?

File: src/ui/tabview/TabPage.qml (hunk 18 above)
Comment (PoisonousJohn) [reply src/ui/tabview/TabPage.qml:30]: It's relevant. It means that activeTab property itself doesn't do anything. It should be a functions assigned by another component. This function should return an activeTab. I did this to avoid dependency on TabController

**Comment by PoisonousJohn** [comment]
> We also probably want to prevent saving the icon url for liri://
I would leave it as is for consistency, since it works fine. I don't want to check url for "liri://".

File: src/ui/window/BrowserWindow.qml [hunk 21]
```diff
@@ -502,7 +502,7 @@ ApplicationWindow {
                     'loadContent': false,
                     'url': restoreTabs[i].url,
                     'title': restoreTabs[i].title,
-                    'iconUrl': restoreTabs[i].iconUrl
+                    'iconUrl': restoreTabs[i].icon.replace("https", "http")
```
Comment (PoisonousJohn) [inline src/ui/window/BrowserWindow.qml]: This bothers me. Right now QML Image is not able to load images via https, showing SSL error in console. Yet browser works fine with SSL. Don't know how to fix it, so it's a hack right now.

Comment (timsueberkrueb) [reply src/ui/window/BrowserWindow.qml]: Interesting, I cannot reproduce this issue here on Ubuntu 17.10 with Qt 5.10.1.

File: src/ui/tabview/TabPage.qml (hunk 18 above)
Comment (timsueberkrueb) [reply src/ui/tabview/TabPage.qml:30]: Ah, all right.

File: src/ui/tabview/content/SettingsContent.qml [hunk 22]
```diff
@@ -61,11 +61,35 @@ TabContent {
...
 
                 FluidControls.TitleLabel {
-                    text: qsTr("Start")
+                    text: qsTr("Startup")
                 }
 
-                Label {
-                    text: qsTr("Start url")
+                Column {
+                    Label { text: "Start new window with" }
```
Comment (timsueberkrueb) [inline src/ui/tabview/content/SettingsContent.qml]: Use `qsTr()`

**Comment by timsueberkrueb** [comment]
Looking good otherwise!

**Comment by PoisonousJohn** [comment]
@tim-sueberkrueb  I believe everything is ok now :)

**Comment by timsueberkrueb** [comment]
@PoisonousJohn I think you need to remove `replace("https", "http")` :)

**Comment by PoisonousJohn** [comment]
@tim-sueberkrueb yeah, sure, I forgot about that, still didn't test it in a "Empty" project

**Comment by PoisonousJohn** [comment]
@tim-sueberkrueb done

**Review by timsueberkrueb — APPROVED** [review]

**Comment by timsueberkrueb** [comment]
@PoisonousJohn do you have any preference on merging the PR? Should I squash your commits?

**Comment by timsueberkrueb** [comment]
Thank you @PoisonousJohn!
